### PR TITLE
added a script that can copy iocs

### PR DESCRIPTION
**Pull request by SamJenkins1**
While working on https://github.com/ISISComputingGroup/IBEX/issues/6898 created a script to more easily duplicate iocs.

### CODE REVIEW

File: ioc_copier/ioc_copier.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+            print("please answer y or n")
+answer = ""
+while answer != "y":
+    answer = input("Have you set the ioc, start_copy, initial_copy, and max_copy variables in the file? y/n: ")
+    if answer == "n":
+        print("open the file and set those variable to match the ioc to copy")
+        exit()
+    else:
+        if answer != "y":
+            print("please answer y or n")
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:27]: Could this be placed into a function so we can possibly add a test in the future if we ever feel the need too while also making the functionality more reusable by following the DRY methodology?

Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:27]: If you agree with my suggestion to set the `ioc`, `start_copy`, `initial_copy` and `max_copy` variables to `None` by default and set as command line arguments for the user to enter on executing hte script, this could be altered to just fail on the variables not being set instead of trusting that the user has set them.

File: ioc_copier/ioc_copier.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                                                                                    f"IOC_{current_copy}")))
+            if f"IOC-{start_copy}" in file:
+                os.rename(os.path.join(root, file), os.path.join(root, file.replace(f"IOC-{start_copy}",
+                                                                                    f"IOC-{current_copy}")))
+        for folder in subFolder:
+            if f"IOC_{start_copy}" in folder:
+                os.rename(os.path.join(root, folder), os.path.join(root, folder.replace(f"IOC_{start_copy}",
+                                                                                        f"IOC_{current_copy}")))
+            if f"IOC-{start_copy}" in folder:
+                os.rename(os.path.join(root, folder), os.path.join(root, folder.replace(f"IOC-{start_copy}",
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:61]: To reduce the level of nested for-loops, do you agree that this might be worth splitting into functions which in turn compartmentalise functionality for easier debugging should we ever need too while also increasing readability of what is going on where each function has a descriptive docstring. This would also allow for simple unit tests to be written to check everything is working as expected when changes are made going forward during the review stage.

File: ioc_copier/ioc_copier.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+                                                                                        f"IOC-{current_copy}")))
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:62]: NITPICK: To make the slight naming different between these as obvious as possible so that no one accidentally removes one from not noticing the difference, I think we should either add verbose comments here for `IOC_{start_copy}` and `IOC-{start_copy}`, and `IOC_{current_copy}` and `IOC-{current_copy}` or place in explicitly named variables.

File: ioc_copier/ioc_copier.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                                                                                    f"IOC_{current_copy}")))
+            if f"IOC-{start_copy}" in file:
+                os.rename(os.path.join(root, file), os.path.join(root, file.replace(f"IOC-{start_copy}",
+                                                                                    f"IOC-{current_copy}")))
+        for folder in subFolder:
+            if f"IOC_{start_copy}" in folder:
+                os.rename(os.path.join(root, folder), os.path.join(root, folder.replace(f"IOC_{start_copy}",
+                                                                                        f"IOC_{current_copy}")))
+            if f"IOC-{start_copy}" in folder:
+                os.rename(os.path.join(root, folder), os.path.join(root, folder.replace(f"IOC-{start_copy}",
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:100]: Similarly to lines 30-61, could this be broken up into functions to compartmentalise functionality a bit more whilst also increasing readability with descriptive docstrings.

File: ioc_copier/ioc_copier.py (hunk 1 above)
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:27]: If you believe that swapping this to command line arguments is not the best solution, in my own personal preference as a user when asked if I have set the `ioc`, `start_copy`, `initial_copy` and `max_copy` variables, I would like the see what has been set returned to me in the terminal to save me opening up the script to check they are set correctly.

File: ioc_copier/ioc_copier.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
+from shutil import copytree
+from pathlib import Path
+import os
+import re
+ioc = "DFKPS"  # The name of the ioc to run, folders to copy will be of the form, <ioc>-IOC-01App and ioc<ioc>-IOC-01
+start_copy = 11  # The folder to copy (avoid IOC1 as other IOCs usually reference this rather than copying it directly)
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:6]: This fails for me when set to an integer lower than 10 02, 03, 04 etc as integers beginning with 0 won't be recognised as integers and if entering 6, a 0 is not placed at the start meaning for the EUROTHRM IOC, when trying to copy 06, the directory is not found.

**Review by JackEAllen — CHANGES_REQUESTED** [review]
Overall, looks good, but sadly does not work for me when testing copying a EUROTHRM 06

When setting the variables to the following:
```Python
ioc = "EUROTHRM"  # The name of the ioc to run, folders to copy will be of the form, <ioc>-IOC-01App and ioc<ioc>-IOC-01
start_copy = 6  # The folder to copy (avoid IOC1 as other IOCs usually reference this rather than copying it directly)
initial_copy = 7  # The first copy
max_copy = 8  # The final copy
``` 
I will receive an error as 0 is not placed in-front of 6 so the correct folder is not located:
```Python
C:\Instrument\Apps\EPICS\ioc\master>python ioc_copier.py
Have you copied ioc copier to EPICS/ioc/master/<ioc_to_copy>? y/n: y
Have you set the ioc, start_copy, initial_copy, and max_copy variables in the file? y/n: y
Traceback (most recent call last):
  File "ioc_copier.py", line 31, in <module>
    copytree(os.path.join(os.getcwd(), f"{ioc_name}-{start_copy}App"),
  File "C:\INSTRU~1\Apps\Python3\lib\shutil.py", line 552, in copytree
    with os.scandir(src) as itr:
FileNotFoundError: [WinError 3] The system cannot find the path specified: 'C:\\Instrument\\Apps\\EPICS\\ioc\\master\\EUROTHRM-IOC-6App'
```

I would like to see a few changes made to increase readability by breaking things up into functions and to make the script PEP8 compliant as currently fails Pylint checks (I've pasted the results I get below)

Completely optionally, I think a few basic tests in a new file would be beneficial to check that the script executes as intended. They could also be used by a reviewer to run as part of the review if changes are ever made going forward. 

When running Pylint, I currently get the following warnings:
```
************* Module ioc_copier
ioc_copier.py:5:0: C0301: Line too long (117/100) (line-too-long)
ioc_copier.py:6:0: C0301: Line too long (119/100) (line-too-long)
ioc_copier.py:21:0: C0301: Line too long (111/100) (line-too-long)
ioc_copier.py:51:0: C0301: Line too long (104/100) (line-too-long)
ioc_copier.py:52:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:54:0: C0301: Line too long (104/100) (line-too-long)
ioc_copier.py:55:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:58:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:59:0: C0301: Line too long (112/100) (line-too-long)
ioc_copier.py:61:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:62:0: C0301: Line too long (112/100) (line-too-long)
ioc_copier.py:83:0: C0301: Line too long (102/100) (line-too-long)
ioc_copier.py:90:0: C0301: Line too long (104/100) (line-too-long)
ioc_copier.py:91:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:93:0: C0301: Line too long (104/100) (line-too-long)
ioc_copier.py:94:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:97:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:98:0: C0301: Line too long (112/100) (line-too-long)
ioc_copier.py:100:0: C0301: Line too long (108/100) (line-too-long)
ioc_copier.py:101:0: C0301: Line too long (112/100) (line-too-long)
ioc_copier.py:102:0: C0303: Trailing whitespace (trailing-whitespace)
ioc_copier.py:103:0: C0301: Line too long (109/100) (line-too-long)
ioc_copier.py:104:0: C0301: Line too long (112/100) (line-too-long)
ioc_copier.py:1:0: C0114: Missing module docstring (missing-module-docstring)
ioc_copier.py:5:0: C0103: Constant name "ioc" doesn't conform to UPPER_CASE naming style (invalid-name)
ioc_copier.py:6:0: C0103: Constant name "start_copy" doesn't conform to UPPER_CASE naming style (invalid-name)
ioc_copier.py:7:0: C0103: Constant name "initial_copy" doesn't conform to UPPER_CASE naming style (invalid-name)
ioc_copier.py:8:0: C0103: Constant name "max_copy" doesn't conform to UPPER_CASE naming style (invalid-name)
ioc_copier.py:10:0: C0103: Constant name "answer" doesn't conform to UPPER_CASE naming style (invalid-name)
ioc_copier.py:15:8: R1722: Consider using sys.exit() (consider-using-sys-exit)
ioc_copier.py:19:0: C0103: Constant name "answer" doesn't conform to UPPER_CASE naming style (invalid-name)
ioc_copier.py:24:8: R1722: Consider using sys.exit() (consider-using-sys-exit)
ioc_copier.py:41:16: C0200: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
ioc_copier.py:76:16: C0200: Consider using enumerate instead of iterating with range and len (consider-using-enumerate)
ioc_copier.py:2:0: W0611: Unused Path imported from pathlib (unused-import)

-----------------------------------
Your code has been rated at 5.88/10
```

The ioc-copier script will certainly be a really nice utility to use going forward so thank you for thinking of the idea and creating it! 🥇

**Review by JackEAllen — COMMENTED** [review]
Overall, the changes are really good! Only have minor changes suggested to the ioc copier to break up the reading and writing to files a little further and over the edge case of trying to copy an existing IOC and to condense tests using parameterized. 👍

File: ioc_copier/ioc_copier.py [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        sys.exit()
+    elif len(sys.argv) > 5:
+        print("Too many arguments")
+        print("Arguments should be <ioc-to-duplicate> <number-ioc-to-copy> "
+              "<first-copy> <max-number-ioc>")
+        print("use argument \"-h\" for more details.")
+        sys.exit()
+    else:
+        ioc = sys.argv[1]
+        global start_copy
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:108]: Constants should conform to upper case naming style and because it's a global variable, should be defined at module level. This will fail pylint checks under `invalid-name` and also `global-variable-undefined`

File: ioc_copier/ioc_copier.py [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+    Function to walk through folders and rename them.
+    :param ioc: The ioc name.
+    :param padded_current_copy:  the current iteration of the ioc number, zero padded
+    :param padded_start_copy: the ioc number to copy, zero padded.
+    :param root:
+    :param sub_folder:
+    :return:
+    """
+    for folder in sub_folder:
+        rename_files(root, folder, ioc, padded_start_copy, padded_current_copy)
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:193]: There are too many positional arguments here for the `rename_files()` function call.

File: ioc_copier/ioc_copier.py [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+def replace_text(text_lines, ioc):
+    """
+    Function to handle replacing of text within files.
+    Parameters:
+        text_lines - the text from the file to process.
+        start - the value to look for to remove.
+        current - the value to replace start with.
+    """
+    for i in range(len(text_lines)):
```
Comment (JackEAllen) [inline ioc_copier/ioc_copier.py:42]: Could this be swapped out to use an enumerate instead of iterating with range and len?

Comment (SamJenkins1) [reply ioc_copier/ioc_copier.py:42]: Pretty sure I tested that and it doesn't work, if I don't use the text_lines[i] style later it doesn't overwrite it within text_lines properly.

File: ioc_copier/ioc_copier.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        max_copy - The final copy.
+        file_format - The format to use for the folder name, either ending
+                      in app or starting with ioc. in the form of an fstring.
+        ioc_name - The name of the ioc to run, folders to copy will be of the form,
+                   <ioc>-IOC-01App and ioc<ioc>-IOC-01.
+    """
+    ioc_name = f"{ioc}-IOC"
+    global current_copy
+    global padded_start_copy
+    global padded_current_copy
```
[the remark this answers is not in the document]
Comment (JackEAllen) [reply ioc_copier/ioc_copier.py:212]: They should be okay to leave as lower case then.

File: ioc_copier/ioc_copier.py (hunk 8 above)
Comment (JackEAllen) [reply ioc_copier/ioc_copier.py:42]: so something like `for i, line in enumerate(text_lines):` wouldn't work?

Comment (SamJenkins1) [reply ioc_copier/ioc_copier.py:42]: I mean, you can use that as the loop condition, but the `line` variable is useless, it doesn't work as an argument to the regex substitution, and it doesn't appear to be a reference to the line, so setting it to the new value doesn't overwrite the version of it actually in the list like doing `text_lines[i] = temp_text` does

Comment (SamJenkins1) [reply ioc_copier/ioc_copier.py:42]: so I can change it to enumerate if you want, but baiscally its going to be `for i, _ in enumerate(text_lines):`

Comment (SamJenkins1) [reply ioc_copier/ioc_copier.py:42]: hmmm, to handle this in a more pythonic way maybe I should pull out the loop internals into there own function and use a list comprehension instead?

**Review by JackEAllen — APPROVED** [review]
I'm sure you'll be happy to here this... LGTM!!! 🥳🎉
